### PR TITLE
Test960 before io freeze + changes to C API for Windows + better output reporting + better user arguments to run locally

### PR DESCRIPTION
**Pull request by jmarrec**
See it in action on my fork: https://travis-ci.com/github/jmarrec/EPTravisTester/builds/235044074

### CODE REVIEW

File: ep_testing/config.py [hunk 1]
```diff
@@ -33,19 +33,25 @@ class OS:
...
         self.bitness = this_config['bitness']
 
-        self.this_version = '9.4'
-        self.tag_this_version = 'v9.4.0'
-        self.last_version = '9.3'
-        self.tag_last_version = 'v9.3.0'
+        self.this_version = '9.6'
+        self.tag_this_version = 'v9-6-0-TestBuild1'
+        self.last_version = '9.5'
+        self.tag_last_version = 'v9.5.0'
```
Comment (jmarrec) [inline ep_testing/config.py:54]: Bump version to test

File: ep_testing/config.py [hunk 2]
```diff
@@ -33,19 +33,25 @@ class OS:
...
 
         # invalid keys are protected in the command's finalize_options method
         this_config = CONFIGURATIONS[run_config_key]
         self.os_version = this_config['os_version']
         self.os = this_config['os']
+        self.msvc_version = None
+        if msvc_version is not None:
+            self.msvc_version = msvc_version
+        elif self.os == OS.Windows:
+            self.msvc_version = 15  # Default to 2017 for Travis
```
Comment (jmarrec) [inline ep_testing/config.py:46]: New keyword, default to MSVC 15 (2017) for travis

File: ep_testing/tester.py [hunk 3]
```diff
@@ -39,10 +39,10 @@ def run(self):
                 self.install_path, self.verbose, {'test_file': '1ZoneUncontrolled.idf', 'binary_sym_link': True}
             )
         TestCAPIAccess().run(
-            self.install_path, self.verbose, {'os': self.config.os, 'bitness': self.config.bitness}
+            self.install_path, self.verbose, {'os': self.config.os, 'bitness': self.config.bitness, 'msvc_version': self.config.msvc_version}
         )
         TestCppAPIDelayedAccess().run(
-            self.install_path, self.verbose, {'os': self.config.os, 'bitness': self.config.bitness}
+            self.install_path, self.verbose, {'os': self.config.os, 'bitness': self.config.bitness, 'msvc_version': self.config.msvc_version}
```
Comment (jmarrec) [inline ep_testing/tester.py:45]: Pass the msvc_version to the C / CPP api tests, used for generator selection

File: ep_testing/tests/api.py [hunk 4]
```diff
@@ -54,44 +63,63 @@ def run(self, install_root: str, verbose: bool, kwargs: dict):
...
+                    command_line.extend(['-G', 'Visual Studio 16 2019', '-A', 'x86'])
+
+            elif msvc_version == 17:
+                if bitness == 'x64':
+                    command_line.extend(['-G', 'Visual Studio 17 2022', '-A', 'x64'])  # default to 64, but be explicit
+                elif bitness == 'x32':
+                    command_line.extend(['-G', 'Visual Studio 17 2022', '-A', 'x86'])
             else:
-                raise EPTestingException('Bad bitness sent to make_build_dir_and_build')
+                raise EPTestingException("Unknown msvc_version passed to make_build_dir_and_build")
```
Comment (jmarrec) [inline ep_testing/tests/api.py:112]: Allow using MSVC 16 (2019) and 17 (2022 preview) in addition to MSVC 15 (2017)

File: ep_testing/tests/api.py [hunk 5]
```diff
@@ -54,44 +63,63 @@ def run(self, install_root: str, verbose: bool, kwargs: dict):
...
-                py = 'C:\\Python36\\Python.exe'
+            # TODO: why do we have to pass the full path to python on darwin/Windows but not linux?
+            # Anyhoo, let's just use sys.executable everywhere
+            # if platform.system() == 'Linux':
+            #     py = 'python3'
+            # elif platform.system() == 'Darwin':
+            #     py = '/usr/local/bin/python3'
+            # else:  # windows
+            #    py = 'C:\\Python36\\Python.exe'
+            py = sys.executable
```
Comment (jmarrec) [inline ep_testing/tests/api.py:74]: Don't harcorde the path to python. Assuming a full path is needed, just use the current full path

File: ep_testing/tests/api.py [hunk 6]
```diff
@@ -17,11 +18,19 @@ def api_resource_dir() -> str:
...
     else:
-        with open(os.devnull, 'w') as dev_null:
-            check_call(command_line, stdout=dev_null, stderr=STDOUT, **kwargs)
+        raise EPTestingException(
+            f'Command {command_line} failed with exit status {r.returncode}!\n'
+            'stderr:\n'
+            f'{r.stderr.decode().strip()}'
+            '\n\n'
+            'stdout:\n'
+            f'{r.stdout.decode().strip()}')
```
Comment (jmarrec) [inline ep_testing/tests/api.py:33]: Improve error reporting

File: ep_testing/tests/api_templates/delayed_cpp_source_windows.cpp [hunk 7]
```diff
@@ -1,22 +1,41 @@
...
+  init = (INITFUNCTYPE)GetProcAddress((HINSTANCE)hInst, "initializeFunctionalAPI");
+  if (!init) {
+    std::cerr << "Cannot get function \\n";
+    return 1;
+  }
+  std::cout << "Calling to initialize via init(state)\\n";
+  init(state);
+  std::cout << "Closing library\\n";
+  FreeLibrary((HINSTANCE)hInst);
+  return 0;
```
Comment (jmarrec) [inline ep_testing/tests/api_templates/delayed_cpp_source_windows.cpp:40]: Need to initialize an EnergyPlusState in order to call `initializeFunctionalAPI`

File: ep_testing/tests/energyplus.py [hunk 8]
```diff
@@ -27,8 +26,18 @@ def run(self, install_root: str, verbose: bool, kwargs: dict):
...
-            raise EPTestingException('EnergyPlus failed!')
+        else:
+            raise EPTestingException(
+                'EnergyPlus failed!\n'
+                f'Command {cmd} failed with exit status {r.returncode}!\n'
+                'stderr:\n'
+                f'{r.stderr.decode().strip()}'
+                '\n\n'
+                'stdout:\n'
+                f'{r.stdout.decode().strip()}')
```
Comment (jmarrec) [inline ep_testing/tests/energyplus.py:43]: Improve error reporting. I may need to also wrap the subprocess.run in a try block to catch "File not Found" and co

File: setup.py [hunk 9]
```diff
@@ -1,40 +1,67 @@
...
-    """A custom command to run E+ tests using `setup.py run --run_config <key>`"""
+    """A custom command to run E+ tests using `setup.py run --run_config <key>`
+
+    Locally it is also possible to path two extract parameters:
+    * the path to where you extracted the tar.gz/zip manually, and,
+    * on windows: the version of MSVC to use
+
+    eg: `python setup.py run --verbose_output --run-config win64 --msvc-version 16 --extracted-install-path "C:\EnergyPlus-9.6.0-ed3a9d36c8-Windows-x86_64"`
+
+    """
```
Comment (jmarrec) [inline setup.py:18]: setup.py is enhanced with new docstring showing the new options

File: setup.py [hunk 10]
```diff
@@ -1,40 +1,67 @@
...
 
     description = 'Run E+ tests on installers for this platform'
     user_options = [
         # The format is (long option, short option, description).
         ('run-config=', None, 'Run configuration, see possible options in config.py'),
+        ('extracted-install-path=', 'x', 'Specifies a path where E+ is already extracted to skip downloading it (should be the path to the root folder that contains energyplus.exe'),
+        ('msvc-version=', None, 'For OS.Windows only, specifies a MSVC generator to use. 15 is default, you can override to 16'),
+        # distutils is already claiming --verbose and setting it as default = 1
+        ('verbose-output', None, 'Enable verbose mode'),
```
Comment (jmarrec) [inline setup.py:27]: New options

File: setup.py [hunk 11]
```diff
@@ -1,40 +1,67 @@
...
-    def __init__(self, dist):
-        super().__init__(dist)
-        self.run_config = None
-
     def initialize_options(self):
-        ...
+        self.run_config = None
+        self.extracted_install_path = None
+        self.msvc_version = None
+        self.verbose_output = None
```
Comment (jmarrec) [inline setup.py:34]: I don't think the super().__init__(dist) was needed, so just do it in initialize_options

File: setup.py [hunk 12]
```diff
@@ -1,40 +1,67 @@
...
+                raise Exception("Parameter --msvc_version doesn't apply not non windows OS")
+            try:
+                self.msvc_version = int(self.msvc_version)
+            except:
+                raise Exception("Parameter --msvc_version should be an int like 15 (2017) or 16 (2019)")
+
+        if self.verbose_output is None:
+            self.verbose_output = False
+        else:
+            self.verbose_output = bool(self.verbose_output)
```
Comment (jmarrec) [inline setup.py:52]: Technically `self.verbose_output = bool(self.verbose_output)` is enough, since bool(None) = False.

File: setup.py [hunk 13]
```diff
@@ -1,40 +1,67 @@
...
-        d = Downloader(c, self.announce)
-        self.announce('EnergyPlus package extracted to: ' + d.extracted_install_path(), level=distutils.log.INFO)
-        t = Tester(c, d.extracted_install_path(), verbose)
+        if self.extracted_install_path is None:
+            d = Downloader(c, self.announce)
+            self.extracted_install_path = d.extracted_install_path()
+            self.announce('EnergyPlus package extracted to: ' + self.extracted_install_path, level=distutils.log.INFO)
+        else:
+            self.announce('Using specified EnergyPlus package extracted at: ' + self.extracted_install_path, level=distutils.log.INFO)
+        t = Tester(c, self.extracted_install_path, self.verbose_output)
```
Comment (jmarrec) [inline setup.py:64]: Run, dealing with the new arguments

**Review by jmarrec — COMMENTED** [review]
@Myoldmopar left a walkthrough of my changes

**Comment by Myoldmopar** [comment]
Merging this so I can continue to add some more stuff to it for the IO freeze.  Thanks for this @jmarrec

**Pull request merged**